src/test/datascience/interactive-common/notebookProvider.unit.test.ts

```
@@ -36,90 +42,72 @@ suite('Data Science - NotebookProvider', () => {
+
+        // Set up experiment manager
+        when(experimentsManager.inExperiment(anything())).thenReturn(false);
+
         notebookProvider = new NotebookProvider(
             instance(fileSystem),
             instance(notebookEditorProvider),
             instance(interactiveWindowProvider),
             instance(disposableRegistry),
-            instance(notebookServerProvider)
+            instance(rawNotebookProvider),
+            instance(jupyterNotebookProvider),
+            instance(configuration),
+            instance(experimentsManager)
         );
     });
 
-    test('NotebookProvider getOrCreateNotebook no server', async () => {
-        when(notebookServerProvider.getOrCreateServer(anything())).thenResolve(undefined);
-
-        const notebook = await notebookProvider.getOrCreateNotebook({ identity: Uri('C:\\\\foo.py') });
-        expect(notebook).to.equal(undefined, 'No server should return no notebook');
-    });
-
-    test('NotebookProvider getOrCreateNotebook server has notebook already', async () => {
-        const notebookServer = createTypeMoq<INotebookServer>('jupyter server');
-        const notebookMock = createTypeMoq<INotebook>('jupyter notebook');
-        notebookServer
-            .setup(s => s.getNotebook(typemoq.It.isAny()))
-            .returns(() => Promise.resolve(notebookMock.object));
-        when(notebookServerProvider.getOrCreateServer(anything())).thenResolve(notebookServer.object);
-
-        const notebook = await notebookProvider.getOrCreateNotebook({ identity: Uri('C:\\\\foo.py') });
-        expect(notebook).to.not.equal(undefined, 'Server should return a notebook');
-    });
-
-    test('NotebookProvider getOrCreateNotebook server does not have notebook already', async () => {
-        const notebookServer = createTypeMoq<INotebookServer>('jupyter server');
+    test('NotebookProvider getOrCreateNotebook jupyter provider has notebook already', async () => {
```

re. our previous talk on unit tests. I did have to update them on a refactor which adds some extra time, but at the same time one of them did catch an issue with my original change to add in jupyterNotebookProvider. So maybe good, maybe bad?

---

Caught the change while running the test, or building the test?

---

I'd say the unit tests we have are useful when building/designing stuff for at least getting the API consistent. After that I've never seen them catch a bug.

---

Caught while updating the tests. There was a bit of an issue now that get and create were called in two calls to the provider that in a case with GetOnly on I was throwing instead of returning undefined if a server was not available already.

For stuff like the low level RawKernel work I think that the unit tests I added were super useful. Without needing to add anything else I could pump in messages and see how all the futures were updating. For stuff like this where it's more passing stuff around...I'm a bit less sure on.

---

> For stuff like the low level RawKernel work I think that the unit tests I added were super useful. Without needing to add anything else I could pump in messages and see how all the futures were updating

Yeah this sounds like what I said. Good for developing the initial software. After that (once you have it working), not quite so useful. You'd likely hit all of the cases with a functional test for these things anyway. At least I hope so. But if you changed the implementation, you'd have to update the unit tests (but not the functional tests). Unless your unit tests were independent of the implementation. That would be a better unit test. That's my biggest beef with them. Too tightly coupled to the implementation to be useful after the initial submit. Works good as a check on your initial ideas when you first write them, but after that just a pain to keep up to date.